### PR TITLE
Improve the dev guide handling

### LINKED ISSUE #4536: Improve the dev guide handling
Having given it some thought I think we can improve how we handle the different versions of the developers guide.

## Current Situation

* Upon branching for a major release we remove all docs in the dev_guide folder, apart from the stability page.
* We add a simple link on the release dev_guide index page to the latest version of the docs where the dev guide is still built.

## Desired Situation

* Stop having to remove the dev_guide source upon branching.
* Automatically redirect readers of the docs to the latest version of the dev guide.

## Changes to be made

* Modify the conf.py file so that the dev guide is excluded from builds for released versions of the docs. This could be achieved by using the [excluded_patterns](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-exclude_patterns) configuration option. **However**, we will still need to build a `dev_tools` link in the release version of the docs so that the link shows up in the sidebar of the documentation.
* Move the stability page out of dev_docs as this should still be built for releases, it probably belongs in the code_ref section instead.
* Activate a redirect on read the docs to automatically redirect people who visit the stable URL to the latest one. (This will have to be done by a project admin one the changes have been merged and backported).

Other notes:

Only the /stable version will redirect to /latest so people who end up on specific version URLs will have to click a link to manually redirect to the latest dev_docs, i.e. mirroring the current situation.

Old versions (older than 2.0) can't be changed so they too will still manually redirect.

### PR DESCRIPTION
**Pull request by shubham00jain**
Fixes #4536.

I haven't changed index.rst using jinja like @Cadair suggested. Please first review if I'm going in the right direction.

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -218,7 +224,12 @@
 
 # -- Stability Page ------------------------------------------------------------
 
-with open('./dev_guide/sunpy_stability.yaml', 'r') as estability:
+if is_development == True:
+    url = './dev_guide/sunpy_stability.yaml'
+else:
+    url = './dev_guide_release/sunpy_stability.yaml'
```
Comment (nabobalis) [inline docs/conf.py:230]: I am not sure this is what we are after. 

The goal is prevent the removal of the dev_guide when we do a release. 

The stability page should be moved wholesale.

Comment (shubham00jain) [reply docs/conf.py:230]: The goal was to automatically remove all the files inside dev_guide (except index.rst)  whenever a stable version is released. So I made 2 folders one for stability release and one for dev release and this url tells the location of the `stability.yaml` file. Is this approach wrong?

Comment (nabobalis) [reply docs/conf.py:230]: Does this work? 

I think we want to move the sunpy_stability page regardless of if its a release or not.

**Comment by nabobalis** [comment]
I think two things need to be done:

1. Move the stability.rst to docs/code_ref and that should be where it lives, no matter if it is a release or development
2. Remove the extra files, those pngs and the extra stability file.

**Comment by shubham00jain** [comment]
Things I've done:
1. The stability files only lives inside code_ref now.
2. Removed extra png images from dev_guide_release. 
3. Updated the index.rst to automatically include whichever version we're on.

Things remaining:
> Activate a redirect on read the docs to automatically redirect people who visit the stable URL to the latest one. (This will have to be done by a project admin one the changes have been merged and backported).

This will have to been done by an admin.

File: docs/dev_guide_release/index.rst [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+*****************
+Developer's Guide
+*****************
+
+Please go <here> for our up to date developer’s guide.
+
+.. toctree::
+   :maxdepth: 2
+
+   ../code_ref/stability
```
Comment (nabobalis) [inline docs/dev_guide_release/index.rst:12]: This does not need to be linked by the dev guide.

Comment (nabobalis) [reply docs/dev_guide_release/index.rst:12]: It does need to be linked by the index file for the code ref.

File: docs/dev_guide/index.rst [hunk 3]
```diff
@@ -26,4 +26,4 @@ This goes over the basics and has links to useful tutorials on git.
    remote_data
    config
    funding
-   stability
+   ../code_ref/stability
```
Comment (nabobalis) [inline docs/dev_guide/index.rst:29]:
```suggestion
```
This should linked in the index file for the code_ref.

File: docs/dev_guide_release/index.rst [hunk 4]
```diff
@@ -0,0 +1,12 @@
+.. _dev_guide:
+
+*****************
+Developer's Guide
+*****************
+
+Please go <here> for our up to date developer’s guide.
```
Comment (nabobalis) [inline docs/dev_guide_release/index.rst:7]: How is here dealt with?

**Comment by nabobalis** [comment]
There are still extra png images.

File: docs/conf.py [hunk 5]
```diff
@@ -115,8 +115,14 @@
 # List of patterns, relative to source directory, that match files and
 # directories to ignore when looking for source files.
 # This pattern also affects html_static_path and html_extra_path.
+
 exclude_patterns = ['_build', 'Thumbs.db', '.DS_Store']
 
+if is_development == True:
```
Comment (nabobalis) [inline docs/conf.py:121]:
```suggestion
if is_development:
```

File: docs/dev_guide_release/index.rst (hunk 2 above)
Comment (nabobalis) [inline docs/dev_guide_release/index.rst:12]:
```suggestion
```

File: docs/index.rst [hunk 6]
```diff
@@ -18,3 +18,13 @@ Documentation :any:`index <genindex>` and :any:`Module <modindex>` index.
...
+
+.. toctree::
+  :maxdepth: 2
+
+  guide/index
+  code_ref/index
+  generated/gallery/index
+  dev_guide_release/index
+  whatsnew/index
+  about
```
Comment (nabobalis) [inline docs/index.rst:30]: Why was this added?

**Comment by nabobalis** [comment]
You now seem to have removed the original images, we kind of want those.

File: docs/dev_guide_release/index.rst (hunk 4 above)
Comment (shubham00jain) [reply docs/dev_guide_release/index.rst:7]: <here> should be linked with the updated dev_guide version (most probably by an admin).

File: docs/conf.py [hunk 7]
```diff
@@ -247,4 +253,4 @@ def setup(app):
     app.connect("source-read", rstjinja)
 
     # The theme conf provides a fix for circle ci redirections
-    fix_circleci(app)
+    fix_circleci(app)
```
Comment (nabobalis) [inline docs/conf.py:256]: This change needs to be reverted.

File: docs/dev_guide_release/index.rst (hunk 4 above)
Comment (nabobalis) [reply docs/dev_guide_release/index.rst:7]: I am not sure I follow.

File: docs/index.rst (hunk 6 above)
Comment (shubham00jain) [reply docs/index.rst:30]: I added this because I thought sphinx will need everything again for the toctree of release version. I just built it and checked, it's unnccesory and only `dev_guide_release/index` needs to be added after `dev_guide/index`. Fixed.

**Comment by nabobalis** [comment]
I think with a bit more tidy, it will be ready.

**Comment by shubham00jain** [comment]
The only thing remaining is linking <here> in dev_guide_release.

File: docs/index.rst [hunk 8]
```diff
@@ -16,5 +16,6 @@ Documentation :any:`index <genindex>` and :any:`Module <modindex>` index.
   code_ref/index
   generated/gallery/index
   dev_guide/index
+  dev_guide_release/index
```
Comment (nabobalis) [inline docs/index.rst:19]: How does this work? If one is excluded at build time, I thought sphinx would error due to a missing reference.

Comment (shubham00jain) [reply docs/index.rst:19]: Sphinx just ignores it. I've built it and checked.

Comment (Cadair) [reply docs/index.rst:19]: This isn't what I had in mind, you should be able to remove this once you get the conditional stuff working in the index.rst. That being said if sphinx wont error for this then it will probably make your template logic simpler.

File: docs/dev_guide_release/index.rst (hunk 4 above)
Comment (Cadair) [reply docs/dev_guide_release/index.rst:7]: This should be the same link as it is now.

File: docs/index.rst (hunk 8 above)
Comment (nabobalis) [reply docs/index.rst:19]: Sphinx does raise a warning, which we try to avoid.

Comment (shubham00jain) [reply docs/index.rst:19]: So, what should be the alternative? Should the whole thing be written using jinga?

Comment (nabobalis) [reply docs/index.rst:19]: Good question, it might be worth trying.

Comment (nabobalis) [reply docs/index.rst:19]: I think only the index file would need to have jinja templating.

File: docs/dev_guide_release/index.rst (hunk 4 above)
Comment (shubham00jain) [reply docs/dev_guide_release/index.rst:7]: Got it. I'll link that.

File: docs/index.rst (hunk 8 above)
Comment (Cadair) [reply docs/index.rst:19]: yeah, you can template the index file so if it's a dev version the toctree is included and if it isn't then just the link is shown

**Comment by shubham00jain** [comment]
@nabobalis @Cadair  I tried using jinja2 template in index.rst and I was not able make it work correctly. Can you tell what is needed to be done?

**Comment by nabobalis** [comment]
> @nabobalis @Cadair I tried using jinja2 template in index.rst and I was not able make it work correctly. Can you tell what is needed to be done?

What did you in the index file?

**Comment by shubham00jain** [comment]
@nabobalis 
```
{% raw %}
.. raw:: html
{% endraw %}

{% from .conf import is_development %}

.. toctree::
  :maxdepth: 2

  guide/index
  code_ref/index
  generated/gallery/index
  {if is_development:}
    dev_guide/index
  {else:}
    dev_guide_release/index
  whatsnew/index
  about
```
I changed the toctree to this. I don't think this is correct. Can I use toctree in jinja2 template or do I have to make the whole ul->li->a loop for every folder?

**Comment by nabobalis** [comment]
As long as it gets templated, I don't see why you would need to do anything else.

**Comment by shubham00jain** [comment]
The if and else statements aren't getting templated and are shown as plain text.

**Comment by nabobalis** [comment]
Will probably need to figure out how to get jina tempting to pick that up page, however we did it for the stability page.

**Comment by shubham00jain** [comment]
@Cadair @nabobalis I have tried many approaches but wasn't able to get the desired jinja templating on `index.rst` file. Can you guys please help?

**Comment by nabobalis** [comment]
> @Cadair @nabobalis I have tried many approaches but wasn't able to get the desired jinja templating on `index.rst` file. Can you guys please help?

Where you able to get the templating to work at all?

**Comment by shubham00jain** [comment]
@nabobalis No. If I copy the exact content as in 'stability.rst' file into `index.rst` file, I get the templated file but I make any changes into it the template goes away.

**Comment by nabobalis** [comment]
Can you commit what you have and I can try to have a look.

**Comment by shubham00jain** [comment]
@nabobalis I tried to use the `.. only::` directive and `ifconfig` extension but they don't seem to work as well.

**Comment by shubham00jain** [comment]
@nabobalis Hi. Did you get a chance to look at my commit?

**Comment by nabobalis** [comment]
Sorry no, I will have a go this evening after work.

**Comment by Cadair** [comment]
I think the templating isn't working because of this: https://github.com/sunpy/sunpy/blob/master/docs/conf.py#L237

**Comment by shubham00jain** [comment]
@nabobalis The templating is working now! 😃  Please tell if I missed anything.

File: docs/conf.py [hunk 9]
```diff
@@ -234,7 +241,7 @@ def rstjinja(app, docname, source):
     if app.builder.format != 'html':
         return
     src = source[0]
-    if "Current status" in src[:20]:
+    if "Current status" or "SunPy Documentation" in src[:20]:
```
Comment (nabobalis) [inline docs/conf.py:244]: "Sunpy documentation" could hit other pages?

File: docs/dev_guide_release/index.rst [hunk 10]
```diff
@@ -0,0 +1,7 @@
+.. _dev_guide:
+
+*****************
+Developer's Guide
+*****************
+
+Please go `here <https://docs.sunpy.org/en/latest/dev_guide/index.html>`_ for our up to date developer’s guide.
```
Comment (nabobalis) [inline docs/dev_guide_release/index.rst:7]:
```suggestion
Please go `here <https://docs.sunpy.org/en/latest/dev_guide/index.html>`__ for our up to date developer’s guide.
```

**Comment by nabobalis** [comment]
Yeah the changes look good.

**Comment by shubham00jain** [comment]
@nabobalis I don't think any other file which has "Sunpy Documentation" is templated using jinja, so we should be fine.

**Comment by nabobalis** [comment]
You will need to fix the issues the failing python_codestyle is finding.

**Comment by shubham00jain** [comment]
@nabobalis Sorry I don't think I follow. What issues?

**Comment by nabobalis** [comment]
> @nabobalis Sorry I don't think I follow. What issues?

The "sunpy.sunpy (python_codestyle [linux]) Failing after 1m — python_codestyle [linux] failed" should pass and you will need to solve the problem it found. 

```bash
All changes made by hooks:
diff --git a/docs/conf.py b/docs/conf.py
index bcf1eac8c..febb67db2 100644
--- a/docs/conf.py
+++ b/docs/conf.py
@@ -254,4 +254,4 @@ def setup(app):
     app.connect("source-read", rstjinja)
 
     # The theme conf provides a fix for circle ci redirections
-    fix_circleci(app)
\ No newline at end of file
+    fix_circleci(app)

```

**Comment by shubham00jain** [comment]
@nabobalis I tried running `pre-commit run --all-files` but I don't understand what's wrong. Is an empty line needed at the end of conf.py ?

**Comment by nabobalis** [comment]
> @nabobalis I tried running `pre-commit run --all-files` but I don't understand what's wrong. Is an empty line needed at the end of conf.py ?

Yes.

**Comment by nabobalis** [comment]
> @nabobalis I tried running `pre-commit run --all-files` but I don't understand what's wrong. Is an empty line needed at the end of conf.py ? 
> 

If you run that pre commit command, it should fix the file for you. I am surprised it didn't do anything.

**Comment by Cadair** [comment]
I think this looks fine as it is, although I have one thought that might make it a little bit clearer. Instead of templating the toctree in `docs/index.rst`, why don't we template `docs/dev_guide/index.rst`? We can have a block like:

```
{if is_development}
current content of file
{else}
link to dev version
{endif}
```

**Comment by shubham00jain** [comment]
@Cadair We can do this but then we won't need to use `exclude_patterns`. Just a simple template in `dev_guide/index.rst` would suffice. So should I make that change?

@nabobalis Whenever I run `pre-commit run --all-files` I get an error that my python/pip version is not update (which I have already done). I have tried to use virtualenv as well. Can you suggest some fix?

**Comment by nabobalis** [comment]
> @Cadair We can do this but then we won't need to use `exclude_patterns`. Just a simple template in `dev_guide/index.rst` would suffice. So should I make that change?

Sure.

> @nabobalis Whenever I run `pre-commit run --all-files` I get an error that my python/pip version is not update (which I have already done). I have tried to use virtualenv as well. Can you suggest some fix?

I need more information to debug this. Can I see an log?

**Comment by shubham00jain** [comment]
@nabobalis 
This is the log:
 https://pastebin.pl/view/22960518

**Comment by nabobalis** [comment]
> @nabobalis
> This is the log:
> [pastebin.pl/view/22960518](https://pastebin.pl/view/22960518)

`ERROR: isort-5.5.2 requires Python '>=3.6,<4.0' but the running Python is 3.5.2 you need to use a more up to date version of pytohn.

**Comment by shubham00jain** [comment]
running python3 --version command gives me python 3.8.5

**Comment by nabobalis** [comment]
> running python3 --version command gives me python 3.8.5

Somewhere along the way, pre-commit is not using the correct version of python and that's probably a virtual environment issue. You want to run the tox build directly.

**Comment by Cadair** [comment]
> @Cadair We can do this but then we won't need to use exclude_patterns. Just a simple template in dev_guide/index.rst would suffice. So should I make that change?

I think we will still need to do this, or else there will be a bunch of rst files which aren't included in a toctree which will cause sphinx to throw warnings.

**Comment by shubham00jain** [comment]
@Cadair Can you tell me the exact structure of the docs folder which you want? I tried to store all the contents in a new `dev/mycontents` and then excluding that but the whole dev_guide folder didn't show. I can try doing that again maybe I missed something.

@nabobalis how to run that ?(This is my first 'good' opensource PR, I'm quite new to this.)

**Comment by Cadair** [comment]
> @Cadair Can you tell me the exact structure of the docs folder which you want? I tried to store all the contents in a new `dev/mycontents` and then excluding that but the whole dev_guide folder didn't show. I can try doing that again maybe I missed something.

I guess I am trying to avoid having a new top level folder in the docs tree. So put all the real content in `dev_guide/content/` then rewrite the toctree in index.rst to reference things in that subfolder, then exclude the subfolder in conf.py

File: docs/conf.py [hunk 11]
```diff
@@ -115,8 +115,12 @@
 # List of patterns, relative to source directory, that match files and
 # directories to ignore when looking for source files.
 # This pattern also affects html_static_path and html_extra_path.
+
 exclude_patterns = ['_build', 'Thumbs.db', '.DS_Store']
 
+if is_development == False:
```
Comment (nabobalis) [inline docs/conf.py:121]:
```suggestion
if not is_development:
```

**Comment by Cadair** [comment]
Something seems to have broken the sphinx build: https://dev.azure.com/sunpy/sunpy/_build/results?buildId=10527&view=logs&j=196c4937-6990-58b5-7d18-234ac2f0888b&t=e790bdf0-5cca-5361-ce0e-7cb2a4cbf89d&l=205

> sphinx.errors.ExtensionError: Handler <function rstjinja at 0x7fe919be1ca0> for event 'source-read' threw an exception (exception: unexpected ':')

**Comment by shubham00jain** [comment]
@Cadair I just cloned the code from sunpy repo and made all the changes again, this time very carefully without adding/removing any unnecessary white-spaces but as soon as I make these changes and try to build to them, I get error in `generated/gallery/computer_vision_techniques/off_limb_enhance ` which I didn't even touch. Can you suggest what might be wrong?

**Comment by nabobalis** [comment]
> @Cadair I just cloned the code from sunpy repo and made all the changes again, this time very carefully without adding/removing any unnecessary white-spaces but as soon as I make these changes and try to build to them, I get error in `generated/gallery/computer_vision_techniques/off_limb_enhance ` which I didn't even touch. Can you suggest what might be wrong?

1. What was the error?
2. You can ignore the gallery and I would suggest not building it when you are testing out the documentation.
   `tox -e build_docs -- -D plot_gallery=False`

**Comment by shubham00jain** [comment]
@nabobalis 

I get this ```Extension error:
Handler <function rstjinja at 0x7f01c2a78b80> for event 'source-read' threw an exception (exception: unexpected ':')
make: *** [Makefile:19: html] Error 2``` when I run `make html`. I don't think a **:** is the problem because the same logic runs when I template `docs/index.rst` instead of `docs/dev_guide/index.rst`.

I tried using ```tox -e build_docs -- -D plot_gallery=False``` but I get `ModuleNotFoundError: No module named 'virtualenv.seed.via_app_data'`. This is the same error I get when I tried to use `tox -e codestyle`. For this, I've tried re-installing virtualenv ,python and pip.

And finally I use VSCode with Microsoft's Default Python Linter/Debugger extension installed.

So my question is, is this happening because of indexing a different file?

**Comment by nabobalis** [comment]
> I get this `Extension error: Handler <function rstjinja at 0x7f01c2a78b80> for event 'source-read' threw an exception (exception: unexpected ':') make: *** [Makefile:19: html] Error 2` when I run `make html`. I don't think a **:** is the problem because the same logic runs when I template `docs/index.rst` instead of `docs/dev_guide/index.rst`.

How about removing text in the first if block and seeing it compiles? 

> I tried using `tox -e build_docs -- -D plot_gallery=False` but I get `ModuleNotFoundError: No module named 'virtualenv.seed.via_app_data'`. This is the same error I get when I tried to use `tox -e codestyle`. For this, I've tried re-installing virtualenv ,python and pip.

I am not sure how to help here but something is differently not setup correctly with the virtual environment. I just can't quite pin it. 

> So my question is, is this happening because of indexing a different file?

This looks to be the case.

**Comment by shubham00jain** [comment]
@Cadair 
I was looking at the error and I have found some problems with putting everything inside `dev_guide/contents` eg. in `config.rst` which was previously inside `dev_guide/` has a **relative link** to another file ` :doc:../guide/customization ` which won't work because we have changed its own path. and sphinx starts giving warning here. There might be some more files like these which I couldn't find which are probably causing the error I'm facing.

So, I would like to suggest my previous approach again i.e using 2 folders. In that way, we won't have to check everything inside every rst file in dev_guide.

**Review by Cadair — APPROVED** [review]

**Review by nabobalis — APPROVED** [review]

**Comment by nabobalis** [comment]
Thanks for working on this @shubham00jain. It was a bit more complex than we expected when we opened the original issue.

**Comment by nabobalis** [comment]
Still see
```
/home/vsts/work/1/s/docs/dev_guide/index.rst:9: WARNING: duplicate label newcomers, other instance in /home/vsts/work/1/s/docs/dev_guide/contents/newcomers.rst
/home/vsts/work/1/s/docs/dev_guide/index.rst:9: WARNING: duplicate label remote_data, other instance in /home/vsts/work/1/s/docs/dev_guide/contents/remote_data.rst
```

File: docs/dev_guide/index.rst [hunk 12]
```diff
@@ -1,9 +1,15 @@
 .. _dev_guide:
+{% if is_development %}
```
Comment (nabobalis) [inline docs/dev_guide/index.rst:2]:
```suggestion
{% if not is_development %}
```
Maybe?

Comment (Cadair) [reply docs/dev_guide/index.rst:2]: yep

**Comment by nabobalis** [comment]
Example failed:
```
Extension error:
Here is a summary of the problems encountered when running the examples

Unexpected failing examples:
/home/vsts/work/1/s/examples/plotting/magnetogram_active_regions.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/vsts/work/1/s/examples/plotting/magnetogram_active_regions.py", line 46, in <module>
    srs_table = srs.read_srs(srs_downloaded_files[0])
  File "/home/vsts/work/1/s/.tox/build_docs/lib/python3.8/site-packages/sunpy/io/special/srs.py", line 32, in read_srs
    with open(filepath) as srs:
FileNotFoundError: [Errno 2] No such file or directory: '/home/vsts/work/1/s/.tox/build_docs/tmp/sunpy/data/20110607SRS.txt'
```
Unrelated to the PR.

**Comment by nabobalis** [comment]
Thanks for the PR @shubham00jain
